### PR TITLE
Backport 2.x: PSA storage format test: algorithms

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Cover algorithm encodings and usage flags in the automatically generated storage format test cases.

The following aspects of the storage format are out of scope of this PR:
* Test performing operations allowed by a key. Currently the test generator does not know which algorithms are compatible with which key types.
* Lifetimes and key identifiers.
* Edge cases that call for manually written test cases.

This PR is against 2.x. Forward-port to 3.0: https://github.com/ARMmbed/mbedtls/pull/4559.

### CODE REVIEW

File: tests/scripts/generate_psa_tests.py [hunk 1]
```diff
@@ -254,13 +270,17 @@ def make_test_case(self, key: StorageKey) -> test_case.TestCase:
...
+            flags = []
             # Some test keys have the RAW_DATA type and attributes that don't
             # necessarily make sense. We do this to validate numerical
             # encodings of the attributes.
             # Raw data keys have no useful exercise anyway so there is no
             # loss of test coverage.
-            exercise = key.type.string != 'PSA_KEY_TYPE_RAW_DATA'
-            extra_arguments = ['1' if exercise else '0']
+            if key.type.string != 'PSA_KEY_TYPE_RAW_DATA':
+                flags.append('TEST_FLAG_EXERCISE')
```
Comment (gilles-peskine-arm) [inline tests/scripts/generate_psa_tests.py:280]: This only does what the key's usage allows. As of 6d8582d64c1de41e78e19c56f5ee191bc3e09b1d keys are generated with the usage `PSA_KEY_USAGE_EXPORT`, so only export is tested, not operations.

Comment (gilles-peskine-arm) [reply tests/scripts/generate_psa_tests.py:280]: But in this pull request, we don't yet know what keys can be used for which algorithm. So that's a job for https://github.com/ARMmbed/mbedtls/pull/4444.

File: scripts/mbedtls_dev/macro_collector.py [hunk 2]
```diff
@@ -301,7 +302,7 @@ def read_file(self, header_file):
             self.read_line(line)
 
 
-class InputsForTest(PSAMacroCollector):
+class InputsForTest(PSAMacroEnumerator):
```
Comment (ronald-cron-arm) [inline scripts/mbedtls_dev/macro_collector.py:322]: Question that could help my review. This has been changed the other way around a few commit before (commit" Use InputsForTest in generate_psa_tests"). Not possible to avoid this back and forth ?

Comment (ronald-cron-arm) [reply scripts/mbedtls_dev/macro_collector.py:322]: I am not asking necessarily to change it but I don't really understand why you have to change this back and forth. Could you elaborate a bit on that ?

**Review by ronald-cron-arm — COMMENTED** [review]
This looks almost good to me. Nice commit structure that I have been able to leverage to ease my code review, thanks.

I have mostly only comments about two commit messages that I don't quite understand.

Regarding completeness I've printed the `algorithms` variable in all_keys_for_algorithms() and compared this list of algorithms with that of the PSA spec. I didn't spot any miss, rather the opposite with algorithms not yet defined in the spec.

Looking at the tests not run with the full configuration, they are only related to algorithms that we don't support yet. One could argue that they have to be run anyway though. Should we remove the dependency on the algorithm being supported, at least for some tests ?

**Comment by gilles-peskine-arm** [comment]
> I didn't spot any miss, rather the opposite with algorithms not yet defined in the spec.

The tests are about Mbed TLS's own backward compatibility, not about compatibility with the PSA specification. So the test data is generated from Mbed TLS, and the PSA specification is irrelevant.

> Looking at the tests not run with the full configuration, they are only related to algorithms that we don't support yet. One could argue that they have to be run anyway though. Should we remove the dependency on the algorithm being supported, at least for some tests ?

Apart from not-supported tests, the test dependencies follow a general principle: all the features used by the test (algorithm, key type, etc.) need to be supported. There is no guarantee that you can create a key whose policy includes an unsupported algorithm. It currently works, but I don't think it's useful to go out of our way to verify it, given that we could change how the library works and then we'd have to change the tests back. Furthermore, a follow-up (#4444) will ensure that keys are exercised after loading whenever possible, and this requires the algorithm to be supported, which would make it more complicated to also maintain tests with relaxed dependencies. I don't see how the additional complexity could be worth it to verify a property that we don't even actually guarantee.

**Comment by ronald-cron-arm** [comment]
> 
> Apart from not-supported tests, the test dependencies follow a general principle: all the features used by the test (algorithm, key type, etc.) need to be supported. There is no guarantee that you can create a key whose policy includes an unsupported algorithm. It currently works, but I don't think it's useful to go out of our way to verify it, given that we could change how the library works and then we'd have to change the tests back. Furthermore, a follow-up (#4444) will ensure that keys are exercised after loading whenever possible, and this requires the algorithm to be supported, which would make it more complicated to also maintain tests with relaxed dependencies. I don't see how the additional complexity could be worth it to verify a property that we don't even actually guarantee.

I understand this but then for algorithms we don't support in the library but have a PSA define like SHA3 we currently don't have any test related to storage and reading of this PSA define value, right? Wouldn't it possible for a defined list of known not supported key algorithms to remove the PSA_WANT_xyz dependency and not do the key exercising tests. At least, we would have some testing for those algorithms.

**Comment by gilles-peskine-arm** [comment]
> At least, we would have some testing for those algorithms.

We will in https://github.com/ARMmbed/mbedtls/pull/4444, which ensures that for every algorithm, there is a test case with `depends_on:!PSA_WANT_ALG_xxx` that ensures that attempting an operation with this algorithm fails. These tests don't load the key (if needed) from storage though.

I don't think we need to test that if you create a key allowing an unsupported algorithm and later load that key, the key will still have the correct metadata. It isn't a promise we actually make.

**Comment by ronald-cron-arm** [comment]
> I don't think we need to test that if you create a key allowing an unsupported algorithm and later load that key, the key will still have the correct metadata. It isn't a promise we actually make.

But isn't it what we are going to go through with SiLabs and SHA3 (or DSA not sure)?  They will support algo we don't support in the library, people will store keys related to those algo and if at some point an update of Mbed TLS is needed on such devices we want to be able to determine if the stored keys will still be loaded properly by the new version or not.

**Comment by gilles-peskine-arm** [comment]
Suppose that Silabs adds support for EDDSA. If they run the unit tests in their configuration, which has EDDSA, there will be test cases for EDDSA.

File: scripts/mbedtls_dev/macro_collector.py [hunk 3]
```diff
@@ -105,6 +105,16 @@ def __init__(self) -> None:
             'tag_length': [],
             'min_tag_length': [],
         } #type: Dict[str, List[str]]
+        self.include_intermediate = False
+
+    def is_internal_name(self, name: str) -> bool:
+        """Whether this is an internal macro. Internal macros will be skipped."""
+        if not self.include_intermediate:
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply scripts/mbedtls_dev/macro_collector.py:116]: I added a comment to the initialization of the field in `__init__`. I don't know what else to write here.

File: scripts/mbedtls_dev/macro_collector.py [hunk 4]
```diff
@@ -301,7 +301,7 @@ def read_file(self, header_file):
             self.read_line(line)
 
 
-class InputsForTest(PSAMacroEnumerator):
+class InputsForTest(PSAMacroCollector):
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply scripts/mbedtls_dev/macro_collector.py]: I've reworded the commit messages to add some explanations. This should make it clearer without having to know the history of the macro collector code.

File: scripts/mbedtls_dev/macro_collector.py [hunk 5]
```diff
@@ -105,6 +105,16 @@ def __init__(self) -> None:
             'tag_length': [],
             'min_tag_length': [],
         } #type: Dict[str, List[str]]
+        self.include_intermediate = False
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply scripts/mbedtls_dev/macro_collector.py:112]: I've completely reworded the commit message which was inaccurate (`psa_generate_tests` is still using `InputsForTests`, what changed is the base class of `InputsForTests`).

I agree that this commit is messy. I don't think I split it right. But each commit was somewhat tested and it would be a nontrivial effort to go back and re-split into meaningful, tested commits. I don't think it's worth it at this point. `macro_collector.py` needs further work to clarify the classes, what they can do and what they're intended for. Completing this work is a long-term background effort that goes beyond the scope of the current pull request.

**Comment by gilles-peskine-arm** [comment]
I pushed a new history https://github.com/gilles-peskine-arm/mbedtls/tree/psa-storage-format-test-algorithms-4 which has the same content, but a modified history:

* The fixup commits are squashed.
* “Use InputsForTest in generate_psa_tests” is reworded.
* “Switch psa_generate_tests to PSAMacroEnumerator” is reworded, now “Expand psa_generate_tests to support constructor arguments”.

**Review by gabor-mezei-arm — DISMISSED** [review]
IMO the macro_collector.py needs a proper refactor/redesign but it is out of scope of this PR. As mentioned in the comments it is good enough to be useful.

**Comment by gilles-peskine-arm** [comment]
> IMO the macro_collector.py needs a proper refactor/redesign but it is out of scope of this PR

I agree. The code in that module has evolved quite a bit from its original purpose. I've been doing some refactoring on the fly but more work is needed. I think the feature set needs to stabilize first, so I prefer to continue working on storage format tests before cleaning up `macro_collector.py`.

**Comment by gilles-peskine-arm** [comment]
I force-pushed the new history I mentioned in https://github.com/ARMmbed/mbedtls/pull/4385#issuecomment-845447512. I'll fix the CI later. I don't know yet whether that will involve a rebase.

**Comment by gilles-peskine-arm** [comment]
The merge into `development_2.x` has out-of-date generated files because this PR modifies `generate_psa_tests.py` and the target branch added key usage flags. I've done another rebase and modified each commit that updated the generated files to make them up-to-date. For comparison, the result of merging `development_2.x` and then running `check-generated-files.sh -u` (https://github.com/gilles-peskine-arm/mbedtls/tree/psa-storage-format-test-algorithms-4-merge) is identical.

**Review by gabor-mezei-arm — APPROVED** [review]

**Comment by ronald-cron-arm** [comment]
>
>I don't think we need to test that if you create a key allowing an unsupported algorithm and later load that key, the key will still have the correct metadata. It isn't a promise we actually make.

Looking back at discussions I don't understand the statement above "It isn't a promise we actually make.". Could you elaborate?

**Comment by gilles-peskine-arm** [comment]
> > I don't think we need to test that if you create a key allowing an unsupported algorithm and later load that key, the key will still have the correct metadata. It isn't a promise we actually make.
>
> Looking back at discussions I don't understand the statement above "It isn't a promise we actually make.". Could you elaborate?

The scenario is:
```
…
psa_set_key_algorithm(&import_attributes, alg);
psa_import_key(&import_attributes, …, &key);
… // possibly restart if the key is persistent
psa_get_key_attributes(key, &stored_attributes);
psa_get_key_algorithm(&stored_attributes);
```
where `alg` is an unsupported algorithm, so attempting any operation with this key will fail. The main question is, will `psa_import_key` succeed? And a secondary question is, if it does, is `psa_key_key_algorithm` guaranteed to return the same value? These are unspecified both in the PSA specification and in Mbed TLS. Currently, Mbed TLS does allow creating the key and `psa_key_key_algorithm` will return the original value, but we could decide to change the behavior if it's convenient, for example if our internal representation starts to diverge from the PSA API.

**Comment by ronald-cron-arm** [comment]
ok, thanks for the additional explanations.

**Review by ronald-cron-arm — APPROVED** [review]
